### PR TITLE
Improve performance of RepositoryIterator and remove a possible next request

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?
The CartRuleLoader already makes this educated guess that there won't be a next result when the current iteration doesn't fill the required criteria limit. Why don't do this in general?

### 2. What does this change do, exactly?
Add an other state to the already stateful RepositoryIterator that knows when to stop.

### 3. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by pweyck** [comment]
Tests are failing

File: src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php [hunk 1]
```diff
@@ -23,6 +23,8 @@ class RepositoryIterator
      * @var Context
      */
     private $context;
+    
+    private $endReached = false;
```
Comment (mitelg) [inline src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: annotation is missing

Comment (JoshuaBehrens) [reply src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: Is it? The initial value should be enough to define the type for analysis tools, IDEs and humans. Or is it some guide that one should follow for your code?

Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: I will not complain, if PHPStan does not complain :smile:

Comment (JoshuaBehrens) [reply src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: Uhm phpstan in my projects does not complain about something like that. But I do php7.4 sooo...

Comment (mitelg) [reply src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: the github action will fail, so no worries :wink: 

but first, fix the code style :smile: as @pweyck already said, this fails

Comment (JoshuaBehrens) [reply src/Core/Framework/DataAbstractionLayer/Dbal/Common/RepositoryIterator.php:30]: I made some code feng shui and freshened up the branch

**Comment by J-Rahe** [comment]
Hey @JoshuaBehrens,
The getTotal doesn’t really work since the TOTAL_COUNT_MODE is set to NONE,
if you would change that, getting the total would probably be slower then just running the query another time.
I will therefore close your PR, still thanks for trying to improve this, and if you think we might have overlooked sth. feel free to open up a new PR.

**Comment by JoshuaBehrens** [comment]
I'll have a look how to combine your feedback and the change 👍

**Comment by JoshuaBehrens** [comment]
I looked into it :) see https://github.com/shopware/shopware/pull/3602

**Pull request closed without merging**